### PR TITLE
Added Common GPU Workflows in Docs

### PR DESCRIPTION
**Pull request by lfenzo**
Following the discussion in #1974, this is a very rough draft of the "Common GPU Workflows", feel free to suggest any changes or additions to this new section in the docs.

I also took the liberty of reformatting the `docs/make.jl` to make the pages and titles a bit easier to read.

### CODE REVIEW

File: docs/src/gpu.md [hunk 1]
```diff
@@ -86,6 +86,64 @@ julia> x |> cpu
...
+
+   gpu_train_loader = Flux.DataLoader((gpu_x, gpu_y), batchsize = 32)
+   ```
+   ```julia
+   gpu_train_loader = Flux.DataLoader((xtrain |> gpu, ytrain |> gpu), batchsize = 32)
+   ```
+   ```julia
+   gpu_train_loader = Flux.DataLoader(gpu.(collect.((xtrain, ytrain))), batchsize = 32)
+   ```
+
```
Comment (ToucheSir) [inline docs/src/gpu.md:123]: Could you add a mention for https://cuda.juliagpu.org/stable/usage/memory/#Batching-iterator here? I can provide a short snippet if needed.

File: docs/src/gpu.md [hunk 2]
```diff
@@ -86,6 +86,64 @@ julia> x |> cpu
...
+   gpu_x = gpu(xtrain)
+   gpu_y = gpu(ytrain)
+
+   gpu_train_loader = Flux.DataLoader((gpu_x, gpu_y), batchsize = 32)
+   ```
+   ```julia
+   gpu_train_loader = Flux.DataLoader((xtrain |> gpu, ytrain |> gpu), batchsize = 32)
+   ```
+   ```julia
+   gpu_train_loader = Flux.DataLoader(gpu.(collect.((xtrain, ytrain))), batchsize = 32)
```
Comment (ToucheSir) [inline docs/src/gpu.md:121]: This incantation is a little overkill, but what does work is:
```suggestion
   gpu_train_loader = Flux.DataLoader(gpu((xtrain, ytrain)), batchsize = 32)
```
Or:
```suggestion
   gpu_train_loader = Flux.DataLoader((xtrain, ytrain) |> gpu, batchsize = 32)
```
`cpu` and `gpu` are smart enough to recurse through tuples and namedtuples.

File: docs/src/gpu.md [hunk 3]
```diff
@@ -86,6 +86,64 @@ julia> x |> cpu
...
+```julia
+using BSON
+# ...
+BSON.@save "./path/to/trained_model.bson" model
+
+# in this approach the cpu-transferred model (referenced by the variable `model`)
+# only exists inside the `let` statement
+let model = cpu(model)
+   BSON.@save "./path/to/trained_model.bson" model
+end
```
Comment (ToucheSir) [inline docs/src/gpu.md:140]:
```suggestion
BSON.@save "./path/to/trained_model.bson" model=cpu(model)
```
Is equivalent and saves on the `let`. It also really ought to be properly documented in BSON.jl. I've created https://github.com/JuliaIO/BSON.jl/issues/111 to track this.

**Review by ToucheSir — COMMENTED** [review]
Thanks a lot! A couple of content-related suggestions from me.

**Comment by lfenzo** [comment]
Thanks @ToucheSir for the reviews! I only left one of them pending, as I wasn't sure how to add an example of that in the context of

**Comment by lfenzo** [comment]
Thanks @ToucheSir for the reviews! I only left the first one pending as I wasn't sure how to add an example of the `CuIterator` in the context.

File: docs/src/gpu.md [hunk 4]
```diff
@@ -86,6 +86,69 @@ julia> x |> cpu
  0.7766742
 ```
 
+## Common GPU Workflows
+
+Some of the common workflows involving the use of GPUs are presented below.
+
+### Transferring Training Data
+
+In order to train the model using the GPU both model and the training data have be transferred to GPU memory. This process can be done with the `gpu` function in two different  ways:
```
Comment (CarloLucibello) [inline docs/src/gpu.md:95]:
```suggestion
In order to train the model using the GPU both model and the training data have to be transferred to GPU memory. This process can be done with the `gpu` function in two different  ways:
```

File: docs/src/gpu.md [hunk 5]
```diff
@@ -86,6 +86,69 @@ julia> x |> cpu
...
+   for epoch in 1:nepochs
+       for (xtrain_batch, ytrain_batch) in train_loader
+           x, y = gpu(xtrain_batch), gpu(ytrain_batch)
+           gradients = gradient(() -> loss(x, y), parameters)
+           Flux.Optimise.update!(optimizer, parameters, gradients)
+       end
+   end
+   ```
+
+1. Transferring all training data to the GPU at once before creating the [DataLoader](@ref) object. This is usually performed for smaller datasets which are sure to fit in the available GPU memory. Some possitilities are:
```
Comment (ToucheSir) [inline docs/src/gpu.md:110]:
```suggestion
2. Transferring all training data to the GPU at once before creating the [DataLoader](@ref) object. This is usually performed for smaller datasets which are sure to fit in the available GPU memory. Some possitilities are:
```

File: docs/src/gpu.md [hunk 6]
```diff
@@ -86,6 +86,69 @@ julia> x |> cpu
...
+   gpu_train_loader = Flux.DataLoader((gpu_x, gpu_y), batchsize = 32)
+   ```
+   ```julia
+   gpu_train_loader = Flux.DataLoader((xtrain |> gpu, ytrain |> gpu), batchsize = 32)
+   ```
+   ```julia
+   gpu_train_loader = Flux.DataLoader((xtrain, ytrain) |> gpu, batchsize = 32)
+   ```
+   Note that both `gpu` and `cpu` are smart enough to recurse through tuples and namedtuples.
+
```
Comment (ToucheSir) [inline docs/src/gpu.md:124]:
```suggestion

3. Wrapping the `DataLoader` in [`CUDA.CuIterator`](https://cuda.juliagpu.org/stable/usage/memory/#Batching-iterator) to efficiently move data to GPU on demand:
```julia
using CUDA: CuIterator
train_loader = Flux.DataLoader((xtrain, ytrain), batchsize = 64, shuffle = true)
# ... model, optimizer and loss definitions
for epoch in 1:nepochs
    for (xtrain_batch, ytrain_batch) in CuIterator(train_loader)
        ...
    end
end
```
Note that this works with a limited number of data types. If `iterate(train_loader)` returns anything other than arrays, approach 1 or 2 is preferred.
```

File: docs/src/gpu.md (hunk 1 above)
Comment (ToucheSir) [reply docs/src/gpu.md:124]: Added something as a suggestion below, see how that reads to you.

File: docs/src/gpu.md [hunk 7]
```diff
@@ -86,6 +86,69 @@ julia> x |> cpu
...
+1. Transferring all training data to the GPU at once before creating the [DataLoader](@ref) object. This is usually performed for smaller datasets which are sure to fit in the available GPU memory. Some possitilities are:
+   ```julia
+   gpu_x = gpu(xtrain)
+   gpu_y = gpu(ytrain)
+
+   gpu_train_loader = Flux.DataLoader((gpu_x, gpu_y), batchsize = 32)
+   ```
+   ```julia
+   gpu_train_loader = Flux.DataLoader((xtrain |> gpu, ytrain |> gpu), batchsize = 32)
```
Comment (CarloLucibello) [inline docs/src/gpu.md:118]: This can be removed as really the same as the one above

File: docs/src/gpu.md (hunk 1 above)
Comment (darsnack) [reply docs/src/gpu.md:124]: Another option is `DataLoader(mapobs(gpu, data), ...)` which will push the data movement invocation into the background thread, though I don't know how that interplays with the actual execution on the GPU.

File: docs/src/gpu.md [hunk 8]
```diff
@@ -86,6 +86,69 @@ julia> x |> cpu
...
+   train_loader = Flux.DataLoader((xtrain, ytrain), batchsize = 64, shuffle = true)
+   # ... model, optimizer and loss definitions
+   for epoch in 1:nepochs
+       for (xtrain_batch, ytrain_batch) in train_loader
+           x, y = gpu(xtrain_batch), gpu(ytrain_batch)
+           gradients = gradient(() -> loss(x, y), parameters)
+           Flux.Optimise.update!(optimizer, parameters, gradients)
+       end
+   end
+   ```
```
[the remark this answers is not in the document]
Comment (lfenzo) [reply docs/src/gpu.md:108]: I just wanted to make sure they would be in the same indentation level as the text in the sub items after rendering. But this is not mandatory at all.

File: docs/src/gpu.md (hunk 6 above)
Comment (lfenzo) [reply docs/src/gpu.md:124]: Thank you! I'll also include your comment on the preferred approach.

File: docs/src/gpu.md (hunk 7 above)
Comment (lfenzo) [reply docs/src/gpu.md:118]: Fair enough! I will then add @darsnack suggestion on `mapsobs`.

File: docs/src/gpu.md [hunk 9]
```diff
@@ -86,6 +86,82 @@ julia> x |> cpu
...
+          # ...
+       end
+   end
+   ```
+
+   Note that this works with a limited number of data types. If `iterate(train_loader)` returns anything other than arrays, approach 1 or 2 is preferred.
+
+### Saving GPU-Trained Models
+
+After the training process is done one must always transfer the trained model back to the `cpu` memory scope before saving it to secundary memory. This can be done, as described in the previous section, with:
```
Comment (ToucheSir) [inline docs/src/gpu.md:140]:
```suggestion
After the training process is done, one must always transfer the trained model back to the `cpu` memory scope before serializing or saving to disk. This can be done, as described in the previous section, with:
```
"secondary memory" is a tad too obscure

**Review by ToucheSir — APPROVED** [review]
This is a great addition, thanks @lfenzo!

**Comment by lfenzo** [comment]
Thanks @ToucheSir, @CarloLucibello and @darsnack for the reviews, comments and suggestions!
